### PR TITLE
Features/periods

### PR DESCRIPTION
**Pull request by jokochems**
Hello everyone,

this one bases on the old PR #353 and reincludes the possibility to model multi-period investments, i. e. investments may not only occur at the start time of the optimization, but are allowed in every period. The implementation is based on the foundation provided by @simnh back in 2017.

It brings along quite a lot of changes. The main ones are:
* Introduce a MultiPeriodModel in oemof.solph.models with two time indices: PERIODS and TIMESTEPS, as well as TIMEINDEX which is a tuple of a period and a corresponding timestep.
* Introduce new MultiPeriodBlocks.
* Introduce routines for lifetime tracking / handling.
* Adjust the objective function and introduce routines for discounting cost values.

The aim was to build the features on top of the existing framework. Hence, existing functionalities should not be affected. For the implementation of a multi-period model, some sacrifices had to be made on generalizability of the multi-period model itself.

Related issues:
- #353 
- #342

Remarks:
I described the changes on the modules on some slides which I stored here: [https://github.com/jokochems/multiperiod_oemof-solph/blob/main/multi-period_modeling_oemof-solph_overview.pdf](https://github.com/jokochems/multiperiod_oemof-solph/blob/main/multi-period_modeling_oemof-solph_overview.pdf)

Any feedback on that is highly appreciated! I think having a working version for thi feature is a large benefit though of course some things are missing to allow a full-featured integration into the framework.

### CODE REVIEW

**Comment by simnh** [comment]
This is so great, I am curios to inspect the idea. Thanks for the great new feature on the road!

**Comment by simnh** [comment]
Is there a basic application incl. dataset? This would be great for testing...Would it be possible to use the OSeMOSYS Atlantis toy example to be implemented for this Expansion modelling? I am currently not aware of the difference of OSeMOSYS implementation and the MultiPeriod Model in solph.

**Comment by jokochems** [comment]
> 
> 
> Is there a basic application incl. dataset? is POMMES already available somewhere to have a look? This would be great for testing...Would it be possible to use the OSeMOSYS Atlantis toy example to be implemented for this Expansion modelling? I am currently not aware of the difference of OSeMOSYS implementation and the MultiPeriod Model in solph.


Hi, to be honest, I don't know the OSeMoSys example you are referring to.
There is a hard-coded toy model included here: https://github.com/jokochems/oemof-solph/blob/features/periods/src/oemof/solph/test_multiperiod
You should be able to flexibly adjust this by uncommenting / commenting lines or adding further components. I left this in order to facilitate playing around and doing unsystematic testing.

We are about to publish the dispatch version of POMMES. We do not yet have the multiperiod feature included into our investment model. So the changes made here were pretty much the basis for doing so. But the routines for reading in data should not differ so much compared to the current status quo with one investment time step only.

The role models for implementation were on the one hand the description of [dynELMod](https://www.diw.de/de/diw_01.c.558131.de/publikationen/data_documentation/2017_0088/dynelmod__a_dynamic_investment_and_dispatch_model_for_the_future_european_electricity_market.html) and on the other hand an [example from Tom Brown](https://nworbmot.org/courses/esm-2019/lectures/notebooks/dynamic_investment.html).

**Comment by p-snft** [comment]
To facilitate the review: It would make sense to merge dev into oemof:features/periods first. Alternatively, to avoid double work, we could close #353 in favour of this PR. This PR would then directly merge into oemof:dev. (@simnh, would you be okay with closing your original PR?)

**Comment by simnh** [comment]
yes thats ok for me!

**Comment by jokochems** [comment]
Thanks for your comments and the decision to use this one in favor of the old #353. 
I'm currently trying to resolve the merge conflicts and include the new stuff from oemof/dev. Since the development took me quite a while and I did not regularly check for updates, unfortunately, this one is a bit behind. Also one of my POMMES co-developers noticed that minor additions and slight changes could make sense which I'll include as well.

**Comment by jokochems** [comment]
In a discussion with @p-snft, we agreed that it would make sense to already align this comprehensive feature to the planned new structure, see #588 and #751 in order to avoid extra work in the upcoming restructuring and to facilitate code reviews.

**Comment by p-snft** [comment]
> In a discussion with @p-snft, we agreed that it would make sense to already align this comprehensive feature to the planned new structure, see #588 and #751 in order to avoid extra work in the upcoming restructuring and to facilitate code reviews.

To help you with this one, I started a quick implementation of the submodule split in #752.

**Comment by jokochems** [comment]
Hey guys,
first of all, thank you all for the valuable feedback on this one at the dev & user meeting! :-)

The further plan to tackle this one would be as follows:
* Touch missing modules.
* Decouple it from oemof.solph and create a new forked repo "oemof.solph_mp" or named somewhat similar, as @uvchik suggested.
* Check whether this can be made the standard in the future, i.e. whether regular oemof.solph models as of today could be represented by MultiPeriodModel(s) with one period only, as proposed by @p-snft and @uvchik.

Besides, I uploaded the slides here and replaced the prior version of the outdated presentation which gives an overview on the changes: https://github.com/jokochems/multiperiod_oemof-solph/blob/main/multi-period_modeling_oemof-solph_overview.pdf

**Comment by jokochems** [comment]
Alright, this one is feature complete now, meaning I touched the remaining modules and components and added multieriod versions. I fixed the failing tests.
I'll have to add some additional tests and check for potential bugs, but then this should be ready for a merge resp. a restructuring to a oemof.solph_mp fork.

**Comment by p-snft** [comment]
Did you check whether one period reproduces the well know behaviour of solph?

**Comment by jokochems** [comment]
> 
> 
> Did you check whether one period reproduces the well know behaviour of solph?

No, it does not (yet). This is due to three reasons which could in principle be all tackled:
* objective_values differ: This could be circumvent by using and if-else condition and checking whether `len(model.PERIODS) == 1`
* some constraint formulations differ slightly: This holds for storages where there is no initial_state resp. `initial_state = 0`. Again, one could check for the length of model.PERIODS.
* By default, the PERIODS are obtained from the year information of the timeindex of the energy_system. This does not necessarily have to be a pd.DatetimeIndex, but you could just pass an integer indexing. This is not yet supported and would require to make the formulation a bit more general. But in principle, this shouldn't be too hard as well.

**Comment by jokochems** [comment]
Alright, here's some info on an intermediate state on this one:
* I included a MultiPeriodNonConvex Flow. So it should now contain all the features of solph as a multi-period variant.
* There is some issue with investment in storages I still have to tackle.
* I added tests for the constraints and am now reaching the minimum test coverage of 90 %. I still want to add tests for uncovered sections, though.

To sum up: Seems like this one is ready for a merge soon resp. to discuss, how to further proceed. There is some minor formatting issue, I'll explain in another comment.

**Comment by jokochems** [comment]
**Proceeding** with this one (once it is finished):

As I see it, there are **three options** for which I provide advantages and drawbacks:
1. Make it a standalone fork of oemof.solph
    * advantage: Code structure would be kind of clean
    * drawback: Maintainability is hard since we'd create double structures
2. Generalize it and make it "the new solph"
    * advantage: We would not have to take care of double structures
    * drawback: Some extra work is required and there are not yet 100 % solutions for everything (see discussion above). Code contrains some if / else checks in order to check the type of the model instance and e.g. return the right constraint blocks.
3. Leave as is and standalone fork of oemof.solph
    * (opportunistic) advantage: My workload doesn't explode any further. ;-)
    * drawback: It brings along also some sort of double structure as putting it into a fork.

Maybe we should have a **discussion** on the options once I put the remainder together and you had a first glance at the changelist.

**Comment by jokochems** [comment]
Here's the minor formatting issue that leads tox checks to fail:
In the constraints package, there are the following lines in the `__init__.py` file
```
from .multiperiodinvestment_limit import (  # noqa: F401
    multiperiodinvestment_limit,
)
```
* tox complains about the formatting of imports and tells me to put the import statement in one line since it is shorter than 79 characters.
* But since I have to add the comment `# noqa: F401` in order to ignore the error when the import is not used, this leads to a line length of 82 characters which in turn leads the PEP 8 check to complain.

As I see it, there are two options:
* Accept that there is one of the both error messages thrown which leads the check to fail.
* Rename the imported routine so everything fits on one line.

What do you think or do you have another solution?

**Comment by jokochems** [comment]
>     * There is some issue with investment in storages I still have to tackle.
Luckily, I was able to fix that one. It was the missing initialization of Investment flows when the invest_relations are set.

>     * I added tests for the constraints and am now reaching the minimum test coverage of 90 %. I still want to add tests for uncovered sections, though.
Test coverage is now about the same as before. It has even slightly increased since I added a few tests for things that were uncovered before.

**Comment by jokochems** [comment]
> 
> 
> **Proceeding** with this one (once it is finished):
> 
> As I see it, there are **three options** for which I provide advantages and drawbacks:
> 
>     1. Make it a standalone fork of oemof.solph
>        
>        * advantage: Code structure would be kind of clean
>        * drawback: Maintainability is hard since we'd create double structures
> 
>     2. Generalize it and make it "the new solph"
>        
>        * advantage: We would not have to take care of double structures
>        * drawback: Some extra work is required and there are not yet 100 % solutions for everything (see discussion above). Code contrains some if / else checks in order to check the type of the model instance and e.g. return the right constraint blocks.
> 
>     3. Leave as is and standalone fork of oemof.solph
>        
>        * (opportunistic) advantage: My workload doesn't explode any further. ;-)
>        * drawback: It brings along also some sort of double structure as putting it into a fork.
> 
> 
> Maybe we should have a **discussion** on the options once I put the remainder together and you had a first glance at the changelist.

Okay, from my point of view, this one is **finished** now. We may add some minor additions (such as mapping "real" lifetimes with periods), but this is not of high priority. Of course, there may also be some small bugs which we not yet detected, but I tried to do a thorough testing for the additions and have been quite busy reading lp-files recently. ;-) 

I think it is **time to decide how to proceed** with this one. I listed the options above. Maybe you could have a look at the major changes and we could discuss and have a vote. As I see it, there has been some slight preference to either option 1 or 2.

Since it brings quite a lot of changes,
* I'll once again update the [presentation](https://github.com/jokochems/multiperiod_oemof-solph/blob/main/multi-period_modeling_oemof-solph_overview.pdf) which gives a good overview,
* Can offer to schedule a video conference where I could walk you through and discuss about the changes and
* You may leave out the review of the new tests I introduced in the first place and check this separately.

I'd be very happy for your feedback! @p-snft, @uvchik and @simnh: I assigned you as reviewers because you seem to be most familiar with either the basic solph-structures that were touched or the feature itself. I'm looking forward to discussing with you and others who'd like to comment and decide on the upcoming process.

**Comment by jokochems** [comment]
I'll retackle this one sometime within the next month. Sorry for the long abscence.

I think, I'll try to go for option 2 from the ones above. 

> 2. Generalize it and make it "the new solph"
>     
>    * advantage: We would not have to take care of double structures
>    * drawback: Some extra work is required and there are not yet 100 % solutions for everything (see discussion above). Code contrains some if / else checks in order to check the type of the model instance and e.g. return the right constraint blocks.  

Creating a new fork seems to cause too much trouble with maintainability and double structures. If this won't succeed, the fork alternative is still a fallback. I'm planning on having it ready before the end of this year (hopefully).

**Comment by jokochems** [comment]
The first actions to re-integrate into ("the new") solph seem quite promising. There is still quite some stuff to do / test, though.
Once done, this will drastically simplify the code basis. Quite a bunch of parallel classes which are still contained in the feature branch as of now will be removed.

**Comment by jokochems** [comment]
#810 replaces this one. I found it easier to do the update of the code basis on a new clean feature branch. I'll close this one.
